[#13] 회원가입 기능 추가

choiminu opened the pull request:
### PR 타입(하나 이상의 PR 타입을 선택해주세요)
- [X] 기능 추가
- [ ] 기능 삭제
- [ ] 버그 수정
- [ ] 문서 수정
- [ ] 코드 리팩토링
- [x] 테스트 코드 추가 및 리팩토링
- [ ] 의존성, 환경 변수, 빌드 관련 코드 업데이트

### 반영 브랜치
- develop

### 이슈
이슈번호 : [#13 ](https://github.com/Maptrackers/IssueMap/issues/13)

### 변경 사항

### UserService에서 회원가입 로직 구현
- 비밀번호 암호화를 위해 PasswordEncoder 적용
- 이메일 중복 검사를 추가하여 중복된 이메일 입력 시 예외 발생
- 회원가입 API 구현

### UserController에 /api/users/signup POST 요청 핸들러 추가
- 회원가입 요청에 대한 성공 및 실패 처리
- 예외 처리 추가

### UserExControllerAdvice를 추가하여 IllegalArgumentException 예외를 처리

### 테스트 코드 작성
- UserControllerTest에 회원가입 성공 및 실패 테스트 추가
- UserServiceTest에 회원가입 로직 단위 테스트 추가


### 테스트 결과
![image](https://github.com/user-attachments/assets/280a1666-3ae2-4f4e-885c-9c349c14679a)

### 개선 계획
- `UserService` 클래스에서 DTO 변환 로직을 별도의 메서드나 클래스로 분리하여 코드의 가독성과 재사용성을 높일 예정입니다.
- 응답 명세가 구체적으로 정해지면, HTTP 상태 코드와 메시지를 명확히 정의하고, 클라이언트와의 통신 규약을 개선할 계획입니다.

choiminu committed:
feat: 회원가입 서비스 구현

- UserService 내 회원가입 로직 추가
- UserSignupRequest, UserSignupResponse DTO 클래스 추가

choiminu committed:
feat: 회원가입 API 구현 및 예외 처리 추가

- UserController 내 /api/users/signup 엔드포인트 추가
- IllegalArgumentException 예외 처리를 위한 ControllerAdvice 추가

choiminu committed:
test: UserController 및 UserService 테스트 추가

- UserControllerTest: 회원가입 성공/실패 테스트
- UserServiceTest: 회원가입 로직 테스트

choiminu committed:
feat: Spring Security 기본 설정 추가

- CSRF 및 권한 관련 기본 설정

Do-oya commented on src/main/java/com/maptracker/issuemap/domain/user/service/UserService.java:
```diff
@@ -0,0 +1,52 @@
+
+        String encodedPassword = passwordEncoder.encode(request.getPassword());
+
+        User userDTO = new User().builder()
```
23번째 줄처럼 메소드 추출하시면 더 읽기 편할거 같아요!

Do-oya replied:
```
User userDTO = new User().builder()
                .id(null)
                .username(request.getUsername())
                .email(request.getEmail())
                .nickname(request.getNickname())
                .password(encodedPassword)
                .role(Role.USER)
                .teams(new ArrayList<>())
                .build();
```

Do-oya commented on src/main/java/com/maptracker/issuemap/domain/user/service/UserService.java:
```diff
@@ -0,0 +1,52 @@
+
+        User user = userRepository.save(userDTO);
+
+        return UserSignupResponse.builder()
```
여기부분도 추출하면 좋을것 같습니다!

Do-oya replied:
```
return UserSignupResponse.builder()
                .username(user.getUsername())
                .email(user.getEmail())
                .build();
```

Do-oya reviewed (approved):
테스트까지 좋습니다~!!👍

choiminu committed:
refactor: User 도메인 코드 리팩토링

- UserExceptionHandler 추가로 예외 처리 로직 개선
  - 예외 처리를 중앙화하여 코드 가독성과 유지보수성 향상
- UserMapper 클래스 추가로 DTO 변환 로직 분리
  - User <-> DTO 변환을 담당하는 클래스로 책임 분리
- UserServiceTest 코드 정리
  - 테스트 코드의 가독성 및 유지보수성을 위해 리팩토링

yohan-kang commented:
exception 처리와 테스트 코드까지 꼼꼼하게 하셔서 이해가 잘되었습니다

ijnim1121 commented:
테스트 코드까지 잘 작성해주셨네요👍
